From ed642f5c561fc60bdf56c1f41ab31c30832b5997 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Thu, 20 Sep 2018 14:47:00 +0300
Subject: [PATCH] [SW-1932] Switch to SHA256 instead of cksum

SHA256 provides better defence than cksum.
Switch to use SHA256 instead of cksum.

Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/board/habanalabs/common/post-image.sh b/board/habanalabs/common/post-image.sh
index f1a8f34..ccbd878 100755
--- a/board/habanalabs/common/post-image.sh
+++ b/board/habanalabs/common/post-image.sh
@@ -71,12 +71,10 @@
 	dd status=none conv=sync bs=1M		if=${UBOOT_BINFILE}			>> ${HL_DEV}-spi-flash.bin.nocrc
 
 	# Prepare checksum
-	CKSUM=`cksum ${HL_DEV}-spi-flash.bin.nocrc | cut -d' ' -f1`
-	CKSUM_BE=`printf "%08x" ${CKSUM}`
-	CKSUM_LE=${CKSUM_BE:6:2}${CKSUM_BE:4:2}${CKSUM_BE:2:2}${CKSUM_BE:0:2}
-	echo ${CKSUM_LE} | xxd -p -r							 > ${HL_DEV}-spi-flash.crc
+	CKSUM=`sha256sum ${HL_DEV}-spi-flash.bin.nocrc | cut -d' ' -f1`
+	echo ${CKSUM} | xxd -p -r							 > ${HL_DEV}-spi-flash.crc
 	dd status=none conv=sync bs=64K		if=${HL_DEV}-spi-flash.bin.nocrc	 > ${HL_DEV}-spi-flash.bin
-	dd status=none conv=sync bs=4		if=${HL_DEV}-spi-flash.crc		>> ${HL_DEV}-spi-flash.bin
+	dd status=none conv=sync bs=32		if=${HL_DEV}-spi-flash.crc		>> ${HL_DEV}-spi-flash.bin
 
 	# Convert bin file into Palladium hex format
 	xxd -p -c1 -g1				${HL_DEV}-spi-flash.bin	 		 > ${HL_DEV}-spi-flash.hex
diff --git a/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn b/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
index 6a0e782..31ce393 100755
--- a/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
+++ b/board/habanalabs/common/rootfs_overlay/usr/sbin/flash_burn
@@ -1,10 +1,10 @@
 #!/bin/sh
 
 if [ $# -eq 0 ]; then
-	echo "Usage: ${0} [-c <cksum>] <image>"
+	echo "Usage: ${0} [-c <sha256>] <image>"
 	echo "Usage: ${0} -s <next_id>"
 	echo "       burn <image> into flash sectors"
-	echo "       -c <cksum>, crc32 for <image> validation"
+	echo "       -c <sha256>, sha256 for <image> validation"
 	echo "       -s <next_id>, set <next_id> as valid image for next boot"
 	exit 0
 else
@@ -47,11 +47,9 @@
 fi
 
 if [ -n "${CKSUM}" ]; then
-	CRC32=`cksum ${INFILE} | cut -d' ' -f1`
-	if ! [ "${CKSUM}" -eq "${CRC32}" ]; then
-		CKSUM_HEX=`printf "0x%x" ${CKSUM}`
-		CRC32_HEX=`printf "0x%x" ${CRC32}`
-		echo "Mismatch with provided checksum [expected ${CKSUM_HEX} while calculated ${CRC32_HEX}]"
+	SHA256=`sha256sum ${INFILE} | cut -d' ' -f1`
+	if [ "${CKSUM}" -ne "${SHA256}" ]; then
+		echo "Mismatch with provided checksum [expected ${CKSUM} while calculated ${SHA256}]"
 		exit 2
 	fi
 fi
